internal/retryhttp/client.go

```diff
@@ -37,6 +41,44 @@ var (
 	DefaultTimeout = time.Second * 30
 	DefaultRetries = 5
 	DefaultClient  = NewClient(DefaultTimeout, DefaultRetries)
+
+	// A regular expression to match the error returned by net/http when the
+	// configured number of redirects is exhausted. This error isn't typed
+	// specifically so we resort to matching on the error string.
+	redirectsErrorRe = regexp.MustCompile(`stopped after \d+ redirects\z`)
+
+	// A regular expression to match the error returned by net/http when the
+	// scheme specified in the URL is invalid. This error isn't typed
+	// specifically so we resort to matching on the error string.
+	schemeErrorRe = regexp.MustCompile(`unsupported protocol scheme`)
+
+	retryableStatusCodes = []int{
+		// 4XX Status codes
+		// The server timed out waiting for the request from client.
+		http.StatusRequestTimeout,
+		// Sometimes the server puts a Retry-After response header
+		// to indicate when the server is available to start processing
+		// request from client.
+		http.StatusTooManyRequests,
+		// The server is unwilling to risk
+		// processing a request that might be replayed.
+		http.StatusTooEarly,
+		// 5XX Status codes
+		// The server is currently unable to handle the request. Without
+		// any other information, the client may retry the request.
+		http.StatusInternalServerError,
```

Comment:
We debated this while filing the story; I don't want to retry on 500 status codes. I don't care what the reasoning out there is that people think retrying on 500 is appropriate, for our platform it is not appropriate. 500 means the server encountered an issue the client cannot resolve, and retrying is just going to needlessly hammer the server.

Reply:
That makes sense and from my research, the consensus is that if we know enough about the server and 500 isn't valid to retry then we shouldn't. If this was something that we explicitly decided then it should have been an AC.